File: packages/papyrus-lang-lsp-server/src/handlers/HoverHandler.ts
import { visitAncestors } from 'papyrus-lang/lib/common/TreeNode';
import { Node, NodeKind } from 'papyrus-lang/lib/parser/Node';
import { SymbolKind } from 'papyrus-lang/lib/symbols/Symbol';
import { CancellationToken, MarkupKind, TextDocumentPositionParams } from 'vscode-languageserver';
import { IDocumentHelpers } from '../DocumentHelpers';
import { buildHoverText } from '../features/Descriptions';
import { Handler } from '../RequestHandlerService';

export class HoverHandler implements Handler<'onHover'> {
    private readonly _documentHelpers: IDocumentHelpers;

    constructor(@IDocumentHelpers documentHelpers: IDocumentHelpers) {
        this._documentHelpers = documentHelpers;
    }

    public handleRequest(params: TextDocumentPositionParams, _cancellationToken: CancellationToken) {
        const nodeAtPosition = this._documentHelpers.getNodeAtPosition(params.textDocument.uri, params.position);

        if (nodeAtPosition) {
            for (const ancestor of visitAncestors<Node>(nodeAtPosition, true)) {
                if (ancestor.kind === NodeKind.Identifier) {
                    const { symbols } = nodeAtPosition.script.scriptFile.program.typeChecker.getSymbolsForIdentifier(
                        ancestor
                    );

                    if (symbols.length > 0) {
                        const symbol = symbols[0];
                        if (symbol.kind === SymbolKind.Intrinsic) {
                            return null;
                        }

                        const text = buildHoverText(
                            symbol,
                            nodeAtPosition.script.scriptFile.program.displayTextEmitter
                        );

                        if (!text) {
                            return null;
                        }

                        return {
                            contents: {
                                kind: MarkupKind.Markdown,
                                value: text,
                            },
                        };
                    }

                    return null;
                }
            }
        }

        return null;
    }
}

File: src/DarkId.Papyrus.Server/Features/HoverHandler.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using DarkId.Papyrus.Common;
using DarkId.Papyrus.LanguageService.Program;
using DarkId.Papyrus.LanguageService.Program.Symbols;
using DarkId.Papyrus.LanguageService.Syntax;
using Microsoft.Extensions.Logging;
using OmniSharp.Extensions.JsonRpc;
using OmniSharp.Extensions.LanguageServer.Protocol.Client.Capabilities;
using OmniSharp.Extensions.LanguageServer.Protocol.Models;
using OmniSharp.Extensions.LanguageServer.Protocol.Server;

namespace DarkId.Papyrus.Server.Features
{
    [Method("textDocument/hover")]
    [Parallel]
    public class HoverHandler : IHoverHandler
    {
        public static MarkupContent DisplayTextToMarkupContent(DisplayText displayText)
        {
            var sb = new StringBuilder();
            sb.AppendLine("```papyrus");
            sb.AppendLine(displayText.Text);
            sb.AppendLine("```");

            if (!string.IsNullOrEmpty(displayText.Documentation))
            {
                sb.AppendLine("___");
                sb.AppendLine(displayText.Documentation);
            }

            return new MarkupContent()
            {
                Kind = MarkupKind.Markdown,
                Value = sb.ToString()
            };
        }

        private readonly ProjectManager _projectManager;
        private readonly ILogger _logger;

        public HoverHandler(ProjectManager projectManager, ILogger<HoverHandler> logger)
        {
            _projectManager = projectManager;
            _logger = logger;
        }

        public Task<Hover> Handle(HoverParams request, CancellationToken cancellationToken)
        {
            try
            {
                var scriptFile = _projectManager.GetScriptForFilePath(request.TextDocument.Uri.ToFilePath());
                if (scriptFile == null)
                {
                    return Task.FromResult<Hover>(null);
                }

                var identifier = scriptFile.Node.GetDescendantNodeOfTypeAtPosition<IdentifierNode>(request.Position.ToPosition());
                if (identifier == null)
                {
                    return Task.FromResult<Hover>(null);
                }

                var symbol = identifier.GetDeclaredOrReferencedSymbol();
                if (symbol == null)
                {
                    return Task.FromResult<Hover>(null);
                }

                var displayText = new DisplayTextEmitter().GetDisplayText(symbol);

                return Task.FromResult(new Hover()
                {
                    Contents = new MarkedStringsOrMarkupContent(DisplayTextToMarkupContent(displayText)),
                    Range = identifier.Range.ToRange()
                });
            }
            catch (Exception e)
            {
                _logger.LogWarning(e, "Error while handling request.");
            }

            return Task.FromResult<Hover>(null);
        }

        public TextDocumentRegistrationOptions GetRegistrationOptions()
        {
            return new TextDocumentRegistrationOptions()
            {
                DocumentSelector = Constants.PapyrusScriptSelector
            };
        }

        public void SetCapability(HoverCapability capability)
        {
            capability.DynamicRegistration = true;
        }
    }
}
